### ISSUE TITLE
Extracting data into MATLAB arrays

### ISSUE THREAD

**Issue by DANAJK** [CCPPETMR/xGadgetron#10]
I have started a Wiki for xGadgetron and a [page on how to get data into MATLAB arrays](https://github.com/CCPPETMR/xGadgetron/wiki/Extracting-Images-into-MATLAB-arrays). I am confused about the handling of complex vs real data. In one of the examples there is a line `complex_ImCont.conversion_to_real(1)` - is this just setting an internal flag?

The output from `image_as_array` is real, whether it is called for a complex or real ImageContainer?

How does the `select` method work?

Any chance of a method that returns a MATLAB complex matrix so that the user doesn't have to loop through all the slices and assemble them into a matrix and sort out real and imaginary slices?

**Comment by evgueni-ovtchinnikov** [comment]
>In one of the examples there is a line complex_ImCont.conversion_to_real(1) - is this just setting an internal flag?
Yes, conversion_to_real sets a flag defining how a complex image is converted to a real one:
1: magnitude computed
2: phase taken
3: real part taken
4: imaginary part taken (as real array)

>The output from image_as_array is real, whether it is called for a complex or real ImageContainer?
Yes it is – perhaps better call it image_as_real_array do you think? In Python I also have image_as_complex_array – need to synchronize Matlab interface now, thanks for spotting this.

>How does the select method work?
It has two arguments: increment (default 1) and offset (default 0). So, for undersampled reconstruction I use select(2) to get images and select(2,1) to get gfactors. This is just a quick fix – ideally should use ImageHeader’s image_series_index, but cannot find its description (which index value stands for what).

>Any chance of a method that returns a MATLAB complex matrix so that the user doesn't have to loop through all the slices and >assemble them into a matrix and sort out real and imaginary slices?
Already in Python, need to synchronize Matlab.

**Comment by KrisThielemans** [comment]
I don't understand the `complex_ImCont.conversion_to_real(1)` "internal flag". Do we need this? I would really like to avoid this.

image_as_array etc should follow the same naming as for PET. Why not simply `as_array()`? I would think that the complex one needs to be the default? In fact, I think the function really should itself figure out what the contained data is, and hence do the appropriate thing.

I don't think we need to provide conversions to real, phase etc. MATLAB and Python will do that for the user.

The` select()` stuff is also way too low-level for the normal user. If Gadgetron/GRAPPA returns gfactors, I wouldn't stuff those in whatever is returned by `get_output()`. Instead, have a method `get_gfactors()`.

In summary, I think we need a single method `as_array()`

**Comment by evgueni-ovtchinnikov** [comment]
I believe too that with MR everything should be complex. No need for that confusing conversion_to_real() then.

In fact, I needed the real stuff simply to have a meaningful ImageProcessor in three-chains demos.

Regarding as_array – it may be ok for images (not sure it is for repeated), but do we really need the whole lot of acquisitions as an array? Normally, they live in a file.

As for select(), I was surprised David managed to dig up this purely internal method – it is not in demos so far as I can see. Actually, if the user opts for getting gfactors (by setting recon.compute_gfactors(True)), then recon.get_output() returns a tuple of images and gfactors.

**Comment by KrisThielemans** [comment]
I don't understand "(not sure it is for repeated)".

it makes sense to have some way of sub-selecting data I guess. could be on the object level though. For a 4D image, we probably need a way to get a 3D Image object out for instance. I don't think this should be in the as_array() function?

**Comment by DANAJK** [comment]
The ‘select’ is in https://github.com/CCPPETMR/xGadgetron/blob/master/examples/undersampled_recon.m
for example:
gfacts = gadgetron.MR_extract_real_images(output.select(2,1));

Now, I’ve just realised that this is on the master branch, but the version on the ‘upstream_update’ branch instead contains the following:

[images, gfacts] = recon.get_output();





gfacts = gfacts.real();



Whilst it’s fine to have multiple mechanisms, we should  recommend a clear method to the users.

We need to think about what the proposed ‘as_array’ returns from the internal data. Users are likely to want data in MATLAB complex form if appropriate, without having to handle separate real and imaginary arrays. Mechanisms for selecting slices, time frames etc would be good, but for now just returning the image data or the g-factor data as specified by the user and as MATLAB matrices (possibly more than 3-dimensional) should be enough.

**Comment by KrisThielemans** [comment]
ok great. Only check the upstream branches or the moment.

```
[images, gfacts] = recon.get_output();
gfacts_data=gfacts.as_array();
```

sounds good enough to me. ok to just return everything for now.
